lib/ex_plasma/transaction.ex

```diff
@@ -1,308 +1,189 @@
+  RLP decodes to structure of RLP-items and then produces an Elixir struct.
 
-  ## Example
+  Assume that it represents a raw transaction if it starts with an integer representing the type.
 
-  iex> txn =
-  ...>  %ExPlasma.Transaction{
-  ...>    inputs: [
-  ...>      %ExPlasma.Output{
-  ...>        output_data: nil,
-  ...>        output_id: %{blknum: 0, oindex: 0, position: 0, txindex: 0},
-  ...>        output_type: nil
-  ...>      }
-  ...>    ],
-  ...>    metadata: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>,
-  ...>    outputs: [
-  ...>      %ExPlasma.Output{
-  ...>        output_data: %{
-  ...>          amount: 1,
-  ...>          output_guard: <<29, 246, 47, 41, 27, 46, 150, 159, 176, 132, 157, 153,
-  ...>            217, 206, 65, 226, 241, 55, 0, 110>>,
-  ...>          token: <<46, 38, 45, 41, 28, 46, 150, 159, 176, 132, 157, 153, 217, 206,
-  ...>            65, 226, 241, 55, 0, 110>>
-  ...>        },
-  ...>        output_id: nil,
-  ...>        output_type: 1
-  ...>      }
-  ...>    ],
-  ...>    sigs: [],
-  ...>    tx_data: <<0>>,
-  ...>    tx_type: 1
-  ...>  }
-  iex> ExPlasma.Transaction.to_rlp(txn)
-  [
-    <<1>>,
-    [<<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>],
-    [
-      [
-        <<1>>,
-        [
-          <<29, 246, 47, 41, 27, 46, 150, 159, 176, 132, 157, 153, 217, 206, 65, 226, 241, 55, 0, 110>>,
-          <<46, 38, 45, 41, 28, 46, 150, 159, 176, 132, 157, 153, 217, 206, 65, 226, 241, 55, 0, 110>>,
-          <<1>>
-        ]
-      ]
-    ],
-    0,
-    <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>
-  ]
+  Only validates that the RLP is structurally correct.
+  Does not perform any other kind of validation, use validate/1 for that.
   """
-  @spec to_rlp(t()) :: list()
-  def to_rlp(%__MODULE__{tx_type: tx_type} = transaction) do
-    transaction |> get_transaction_type(tx_type).to_rlp() |> remove_empty_sigs()
+  @spec to_map(list()) :: {:ok, Protocol.t()} | {:error, mapping_error()}
+  def to_map([raw_tx_type | raw_tx_rlp_decoded_chunks]) do
+    case parse_tx_type(raw_tx_type) do
+      {:ok, tx_type} ->
+        protocol_module = @tx_types_modules[tx_type]
+        Protocol.to_map(protocol_module.__struct__, [raw_tx_type | raw_tx_rlp_decoded_chunks])
+
+      {:error, :unrecognized_transaction_type} = error ->
+        error
+    end
   end
 
-  # NB: We need to standardize on this. Currently, if there is no sig, we strip the empty list.
-  defp remove_empty_sigs([[] | raw_transaction_rlp]), do: raw_transaction_rlp
-  defp remove_empty_sigs(raw_transaction_rlp), do: raw_transaction_rlp
+  def to_map(_), do: {:error, :malformed_transaction}
+
+  defp parse_tx_type(tx_type_rlp) do
+    case RlpDecoder.parse_uint256(tx_type_rlp) do
+      {:ok, tx_type} when tx_type in @tx_types -> {:ok, tx_type}
+      _ -> {:error, :unrecognized_transaction_type}
```

Comment:
on line 57 you match the error and here you don't. why?

Reply:
I don't want to bother with all possible error returned by `parse_uint256`, if there is an error when parsing it, we just return an `unrecognized_transaction_type`.
Line 57 is calling `decode` which has only 1 error that can be returned and it fits the context, so we don't need to do anything.